src/latch_cli/services/launch/launch_v2.py

```diff
@@ -19,13 +23,126 @@
 from latch_sdk_config.latch import config
 from latch_sdk_gql.execute import execute
 
+ExecutionStatus = typing.Literal[
+    "UNDEFINED",
+    "QUEUED",
+    "RUNNING",
+    "SUCCEEDED",
+    "ABORTED",
+    "FAILED",
+    "INITIALIZING",
+    "WAITING_FOR_RESOURCES",
+    "SKIPPED",
+    "ABORTING",
+]
+
+
+class Execution:
+    def __init__(self, execution_id: str, python_outputs: dict[str, type]) -> None:
+        self.execution_id = execution_id
+        self.python_outputs = python_outputs
+
+        self.status: ExecutionStatus = "UNDEFINED"
+
+        self.output: Union[dict[str, Any], None] = None
+        self.outputs_url: Union[str, None] = None
+
+    @property
+    def id(self) -> str:
+        return self.execution_id
+
+    def poll_status(self) -> None:
+        while self.status != "SUCCEEDED" or self.outputs_url is None:
+            res: dict[str, Any] = execute(
+                gql.gql(
+                    """
+                    query GetExecutionStatus($executionId: BigInt!) {
+                        executionInfo(id: $executionId) {
+                            id
+                            status
+                            outputsUrl
+                        }
+                    }
+                    """
+                ),
+                {"executionId": self.execution_id},
+            )
+            execution_info = res.get("executionInfo", {})
+            self.status = execution_info.get("status")
+            self.outputs_url = execution_info.get("outputsUrl")
+            time.sleep(1)
+
+    def get_outputs(self) -> dict[str, Any]:
+        if self.output is not None:
+            return self.output
+
+        if self.status != "SUCCEEDED" or self.outputs_url is None:
+            raise ValueError("workflow non successful")
```

Comment:
have a separate error message for if the execution is in progress

Reply:
the API here is kind of weird tbh cause you have to fetch before this function becomes usable

the most correct thing would be for `.wait` to return a new `CompletedExecution` class that would have getters for everything

---
alternatively

```suggestion
        if self.status == "UNDEFINED":
            raise ExecutionNotAwaitedError
        elif self.status != "SUCCEEDED":
            raise ExecutionFailedError("no output structure available. Some file outputs might still be accessible using `.fetch_written_files`")
```